synapse/handlers/e2e_keys.py

```diff
@@ -121,7 +122,54 @@ def query_devices(self, query_body, timeout):
         # Now fetch any devices that we don't have in our cache
         @defer.inlineCallbacks
         def do_remote_query(destination):
+            """This is called when we are querying the device list of a user on
+            a remote homeserver and their device list is not in the device list
+            cache. If we share a room with this user and we're not querying for
+            specific user we will update the cache
+            with their device list."""
+
             destination_query = remote_queries_not_in_cache[destination]
+
+            # We first consider whether we wish to update the device list cache with
+            # the users device list. We want to track a user's devices when the
+            # authenticated user shares a room with the queried user and the query
+            # has not specified a particular device.
+            # If we update the cache for the queried user we remove them from further
+            # queries. We use the more efficient batched query_client_keys for all
+            # remaining users
+            user_ids_updated = []
+            for (user_id, device_list) in destination_query.items():
+                if user_id not in user_ids_updated:
+                    try:
+                        if not device_list:
```

Comment:
For sanity its probably a good idea to reduce some of this indentation through judicious use of `continue`:

```python
for user_id, device_list in destination_query.items():
    if user_id in user_ids_updated:
        continue

    if device_list:
        continue

    room_ids = yield self.store.get_rooms_for_user(user_id)
    if not room_ids:
        continue

    ...
```